Mage.Sets/src/mage/cards/t/TomBombadil.java

```diff
@@ -0,0 +1,188 @@
+        // ability)
+        Ability ability = ability_opt.get();
+        if (!(ability instanceof TriggeredAbility))
+            return false;
+
+        // Make sure it was a chapter ability
+        TriggeredAbility triggeredAbility = (TriggeredAbility) ability;
+        if (!SagaAbility.isChapterAbility(triggeredAbility))
+            return false;
+
+        // There's a chance that the permanent that this abiltiy came from no longer
+        // exists, so try and find it on the battlefield or check last known
+        // information.
+        // This permanent is needed to check if the resolving ability was the final
+        // chapter ability on that permanent.
+        Permanent permanent = game.getPermanentOrLKIBattlefield(event.getSourceId());
+        if (permanent == null
+                || !permanent.isControlledBy(getControllerId())
+                || !permanent.hasSubtype(SubType.SAGA, game)) {
+            return false;
+        }
+        // Find the max chapter number from that permanent
+        int maxChapter = CardUtil
+                .castStream(permanent.getAbilities(game).stream(), SagaAbility.class)
+                .map(SagaAbility::getMaxChapter)
+                .mapToInt(SagaChapter::getNumber)
+                .sum();
+        // Check if the ability was the last one
+        return SagaAbility.isFinalAbility(triggeredAbility, maxChapter);
+    }
+
+    @Override
+    public String getRule() {
+        return "Whenever the final chapter ability of a Saga you control resolves, reveal cards from the top of your library until you reveal a Saga card. Put that card onto the battlefield and the rest on the bottom of your library in a random order.";
+    }
+}
+
+// From PrismaticBridgeEffect
+class TomBombadilEffect extends OneShotEffect {
+
+    public TomBombadilEffect() {
+        super(Outcome.PutCardInPlay);
+        this.staticText = "reveal cards from the top of your library until you reveal a Saga card. Put that card onto the battlefield and the rest on the bottom of your library in a random order";
+    }
+
+    private TomBombadilEffect(final TomBombadilEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public TomBombadilEffect copy() {
+        return new TomBombadilEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        Cards toReveal = new CardsImpl();
+        Card toBattlefield = null;
+        for (Card card : controller.getLibrary().getCards(game)) {
+            toReveal.add(card);
+            if (card.hasSubtype(SubType.SAGA, game)) {
+                toBattlefield = card;
+                break;
+            }
+        }
+        controller.revealCards(source, toReveal, game);
+        if (toBattlefield != null) {
+            toReveal.remove(toBattlefield);
```

Comment:
instead of removing the card here, move it to the battlefield and then use `toReveal.retainZone(Zone.Library, game)` so you don't have to worry about whether it actually went to the battlefield or some other effect prevented it

Reply:
Okay, sounds good. I copy-pasted this wholesale from [`PrismaticBridgeEffect`](https://github.com/magefree/mage/blob/b1b0cd7d40bce897937a5a4ec04a859a63a0fa31/Mage.Sets/src/mage/cards/e/EsikaGodOfTheTree.java#L111), is this something that should change there too?

Reply:
yeah, although it should really all be refactored to use a common class and use the `PutCards` enum which is outside the scope of this PR